package/MDAnalysis/coordinates/TRC.py

```diff
@@ -0,0 +1,399 @@
+
+.. Note::
+   MDAnalysis requires the blocks to be in the same order for each frame
+   and ignores non-supported blocks.
+
+
+
+Classes
+-------
+
+.. autoclass:: TRCReader
+   :members:
+
+"""
+
+import os
+import errno
+import warnings
+import numpy as np
+
+from . import base
+from .timestep import Timestep
+from ..lib import util
+from ..lib.util import cached, store_init_arguments
+from ..exceptions import NoDataError
+from ..version import __version__
+
+import logging
+logger = logging.getLogger("MDAnalysis.coordinates.GROMOS11")
+
+
+class TRCReader(base.ReaderBase):
+    """Coordinate reader for the GROMOS11 format
+    """
+
+    format = 'TRC'
+    units = {'time': 'ps', 'length': 'nm'}
+    _Timestep = Timestep
+
+    @store_init_arguments
+    def __init__(self, filename, **kwargs):
+        super(TRCReader, self).__init__(filename, **kwargs)
+
+        self.SUPPORTED_BLOCKS = ['TITLE', 'TIMESTEP', 'POSITIONRED',
+                                 'GENBOX']
+        self.NOT_SUPPORTED_BLOCKNAMES = ['POSITION', 'REFPOSITION',
+                                         'VELOCITY', 'VELOCITYRED',
+                                         'FREEFORCE', 'FREEFORCERED',
+                                         'CONSFORCE', 'CONSFORCERED',
+                                         'SHAKEFAILPOSITION',
+                                         'SHAKEFAILPREVPOSITION',
+                                         'LATTICESHIFTS',
+                                         'COSDISPLACEMENTS',
+                                         'STOCHINT', 'NHCVARIABLES',
+                                         'ROTTRANSREFPOS', 'PERTDATA',
+                                         'DISRESEXPAVE',
+                                         'JVALUERESEXPAVE',
+                                         'JVALUERESEPS',
+                                         'JVALUEPERSCALE',
+                                         'ORDERPARAMRESEXPAVE',
+                                         'XRAYRESEXPAVE', 'LEUSBIAS',
+                                         'LEUSBIASBAS', 'ENERGY03',
+                                         'VOLUMEPRESSURE03',
+                                         'FREEENERGYDERIVS03',
+                                         'BFACTOR', 'AEDSS']
+
+        # GROMOS11 trajectories are usually either *.trc or *.trc.gz.
+        # (trj suffix can come up when trajectory obtained from clustering)
+        root, ext = os.path.splitext(self.filename)
+        self.trcfile = util.anyopen(self.filename)
+        if (ext[1:] == "trc") or (ext[1:] == "trj"):
+            self.compression = None
+        else:
+            self.compression = ext[1:]
+
+        # Read and calculate some information about the trajectory
+        self.traj_properties = self._read_traj_properties()
+
+        self._cache = {}
+        self.ts = self._Timestep(self.n_atoms, **self._ts_kwargs)
+
+        self._reopen()
+        self.ts.dt = self.traj_properties["dt"]
+
```

Comment:
Normally we set frame to -1 and then call self._read_next_timestep()

Reply:
`self._reopen()` calls `self.open_trajectory()` which sets `self._frame` to -1
`self._read_frame(0)` seeks the beginning of the file and then calls `self._read_next_timestep()`

I tried to look through existing readers and every reader does it a bit differently

My bigger problem is that `self._frame` is used to store current frame number (0-based) (which is used to set `self.ts.frame` in `_frame_to_ts(self, frameDat, ts)`) while `self.ts._frame` is a different (discontinued??) attribute equal to `self.ts.data['step']`. That is super confusing and leads to easy mistakes but seems to be how it is done, also in other readers. 

In documentation 6.1.5.2.2. ts._frame seems to be an attribute equal to  `self.ts.data['step']` while in doc 6.29.1. it does not show up anymore as attribute.

Please help me understand better what is going on and if I got something wrong. I believe that my implementation yields correct results.

Reply:
Yeah, each Reader can (and often do) work differently.  It's a product that they've been written by many different folk at many different times in the project.  The important thing is to make the external facets align...

`self._frame` (where self is the Reader) is arbitrary, so you can (for example) use this for internal book keeping to keep track of what frame the Reader file handle is currently positioned at.  This is probably the pattern you've seen

`ts.frame` is the 0 based sequential count of the frames the Reader produced.  This is the n-th frame in the trajectory, rather than the n-th integrator iteration of the MD engine, so `ts.frame` should be sequential from 0..n-1.

Because `ts.frame` can get modified by users, I wouldn't rely on this to know the internal state of the Reader (see book-keeping `Reader._frame` usage above).

I'm not certain on `ts._frame`, but I wouldn't trust it for the above reasons.

`ts.data` is a dictionary where you can put whatever data doesn't fit into the existing fields.  I think `data['step']` is often used to store the integrator timestep (i.e. if you're saving every 250 steps, this would go 0, 250, 500 etc).  There's no rules here, just each Reader should document what it has placed in this dictionary.